crates/rpc/src/v1/impls/eth.rs

```diff
@@ -696,6 +696,208 @@ where
+        mut block_count: U256,
+        newest_block: BlockNumber,
+        reward_percentiles: Option<Vec<f64>>,
+    ) -> BoxFuture<EthFeeHistory> {
+        let mut result = EthFeeHistory::default();
+
+        if block_count < 1.into() {
+            return Box::new(future::done(Ok(result)));
+        }
+
+        if block_count > 1024.into() {
+            block_count = 1024.into();
+        }
+
+        let latest_block = self.client.chain_info().best_block_number;
+        let pending_block = self.client.chain_info().best_block_number + 1;
+
+        let last_block = match newest_block {
+            BlockNumber::Hash {
+                hash: _,
+                require_canonical: _,
+            } => 0,
+            BlockNumber::Num(number) => {
+                if number <= pending_block {
+                    number
+                } else {
+                    0
+                }
+            }
+            BlockNumber::Latest => latest_block,
+            BlockNumber::Earliest => 0,
+            BlockNumber::Pending => pending_block,
+        };
+
+        let first_block = if last_block >= block_count.as_u64() - 1 {
+            last_block - (block_count.as_u64() - 1)
+        } else {
+            0
+        };
+
+        result.oldest_block = BlockNumber::Num(first_block);
+
+        let get_block_header = |i| {
+            self.client
+                .block_header(BlockId::Number(i))
+                .ok_or_else(errors::state_pruned)
+                .and_then(|h| {
+                    h.decode(self.client.engine().params().eip1559_transition)
+                        .map_err(errors::decode)
+                })
+        };
+
+        let calculate_base_fee = |h| {
+            self.client
+                .engine()
+                .calculate_base_fee(&h)
+                .unwrap_or_default()
+        };
+
+        let calculate_gas_used_ratio = |h: &Header| {
+            let gas_used = match self.client.block_receipts(&h.hash()) {
+                Some(receipts) => receipts
+                    .receipts
+                    .last()
+                    .map_or(U256::zero(), |r| r.gas_used),
+                None => 0.into(),
+            };
+
+            (gas_used.as_u64() as f64) / (h.gas_limit().as_u64() as f64)
+        };
+
+        let get_block_transactions = |i| match self.client.block_body(BlockId::Number(i)) {
+            Some(body) => Some(body.transactions()),
+            None => None,
+        };
+
+        let reward_percentiles = reward_percentiles.unwrap_or_default();
+        let mut reward_final = vec![];
+
+        for i in first_block..=last_block + 1 {
+            let is_last = i == last_block + 1;
+
+            if i < pending_block {
+                match get_block_header(i) {
+                    Ok(h) => {
+                        let base_fee = h.base_fee();
+
+                        result.base_fee_per_gas.push(base_fee.unwrap_or_default());
+
+                        if !is_last {
```

Comment:
Nitpick: I don't think this `if` block is necessary here (Since `i < pending_block <= last_block`, `is_last` cannot be `true`).

Reply:
Actually, it is possible for last_block to be far smaller then pending_block (please check how the last_block is calculated, it can be built from some block number deep in history).